File: VBNetCore.Mvvm/ViewModels/PopupViewModel.vb
Imports VBNetCore.Mvvm.Commands
Imports VBNetCore.Mvvm.Controls
Imports System
Imports System.Windows.Input

Namespace VBNetCore.Mvvm.ViewModels
    Public MustInherit Class PopupViewModel
        Inherits ValidationPropertyChangedBase
        Public Event CloseHandler As EventHandler
        Public Sub New()
            CloseCommand = New RelayCommand(AddressOf OnClose)
        End Sub
        Protected Sub OnClose()
            RaiseEvent CloseHandler(Me, EventArgs.Empty)
        End Sub
        Public Property Result As PopupResult
            Get
                Return GetProperty(Of PopupResult)()
            End Get
            Protected Set(ByVal value As PopupResult)
                SetProperty(value)
            End Set
        End Property
        Public Property Caption As String
            Get
                Return GetProperty(Of String)()
            End Get
            Set(ByVal value As String)
                SetProperty(value)
            End Set
        End Property
        Public ReadOnly Property CloseCommand As ICommand
    End Class
End Namespace

File: VBNetCore.Mvvm/ViewModels/PopupViewModel.cs
using NetCore.Mvvm.Abstractions;
using NetCore.Mvvm.Commands;
using NetCore.Mvvm.Controls;
using System;
using System.Windows.Input;

namespace NetCore.Mvvm.ViewModels
{
    public abstract class PopupViewModel : ValidationPropertyChangedBase
    {
        public event EventHandler? CloseHandler;
        public PopupViewModel()
        {
            CloseCommand = new RelayCommand(OnClose);
        }
        protected void OnClose()
        {
            CloseHandler?.Invoke(this, EventArgs.Empty);
        }
        public PopupResult Result
        {
            get => GetProperty<PopupResult>();
            protected set => SetProperty(value);
        }
        public string Caption
        {
            get => GetProperty<string>();
            set => SetProperty(value);
        }
        public ICommand CloseCommand { get; }
    }
}
